casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -297,103 +297,112 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics] private[rholang] (
       EvaluateResult(cost, errors) = evaluateResult
-      _                            <- span.mark("before-process-deploy-create-checkpoint")
-      newCheckpoint                <- runtime.space.createCheckpoint()
+      _                            <- span.mark("before-process-deploy-create-soft-checkpoint")
+      checkpoint                   <- runtime.space.createSoftCheckpoint()
       deployResult = InternalProcessedDeploy(
         deploy,
         Cost.toProto(cost),
-        newCheckpoint.log,
+        checkpoint.log,
         Seq.empty[trace.Event],
         DeployStatus.fromErrors(errors)
       )
-      newHash = if (errors.isEmpty) newCheckpoint.root else startHash
-      _       <- span.mark("process-deploy-finished")
-    } yield (newHash, deployResult)
+      _ <- if (errors.nonEmpty) runtime.space.revertToSoftCheckpoint(fallback)
+          else Applicative[F].unit
+      _ <- span.mark("process-deploy-finished")
+    } yield deployResult
 
   private def replayDeploys(
+      runtime: Runtime[F],
       startHash: StateHash,
       terms: Seq[InternalProcessedDeploy],
       replayDeploy: (
-          Blake2b256Hash,
           InternalProcessedDeploy
-      ) => F[Either[ReplayFailure, Blake2b256Hash]]
-  ): F[Either[ReplayFailure, StateHash]] = {
-
-    val startHashBlake = Blake2b256Hash.fromByteString(startHash)
-    val result = terms.toList.foldM(startHashBlake.asRight[ReplayFailure]) {
-      case (hash, deploy) =>
-        hash.flatTraverse(replayDeploy(_, deploy))
-    }
-    result.nested.map(_.toByteString).value
-  }
+      ) => F[Either[ReplayFailure, Unit]]
+  ): F[Either[ReplayFailure, StateHash]] =
+    for {
+      _ <- runtime.replaySpace.reset(Blake2b256Hash.fromByteString(startHash))
+      result <- terms.toList.foldM(().asRight[ReplayFailure]) {
+                 case (unit, deploy) =>
+                   unit.flatTraverse { Unit =>
+                     replayDeploy(deploy)
+                   }
+               }
+      res <- EitherT
+              .fromEither[F](result)
+              .flatMapF { _ =>
+                runtime.replaySpace
+                  .createCheckpoint()
+                  .map(finalCheckpoint => finalCheckpoint.root.toByteString.asRight[ReplayFailure])
+              }
+              .value
+    } yield res
 
   private def replayDeploy(runtime: Runtime[F], span: Span[F])(
-      hash: Blake2b256Hash,
       processedDeploy: InternalProcessedDeploy
-  ): F[Either[ReplayFailure, Blake2b256Hash]] = {
+  ): F[Either[ReplayFailure, Unit]] = {
     import processedDeploy._
     for {
       _                    <- span.mark("before-replay-deploy-reset-rig")
-      _                    <- runtime.replaySpace.resetAndRig(hash, processedDeploy.deployLog)
+      _                    <- runtime.replaySpace.rig(processedDeploy.deployLog)
+      softCheckpoint       <- runtime.replaySpace.createSoftCheckpoint()
```

Comment:
What is the purpose of this softCheckpoint? Can we not revert using the passed in processedDeploy somehow?

Reply:
1) We need to be able to revert the changes to the hot store applied by a faulty deploy, so we need to do a snapshot of the state from before those changes
2) I'm not sure what you mean, can you be more specific?